scwconfig/config.go

```diff
@@ -0,0 +1,323 @@
+	terraformSecretKeyEnv    = "SCALEWAY_TOKEN"
+	terraformOrganizationEnv = "SCALEWAY_ORGANIZATION"
+	terraformRegionEnv       = "SCALEWAY_REGION"
+	cliTLSVerifyEnv          = "SCW_TLSVERIFY"
+	cliOrganizationEnv       = "SCW_ORGANIZATION"
+	cliRegionEnv             = "SCW_REGION"
+	cliSecretKeyEnv          = "SCW_TOKEN"
+
+	// TBD
+	cliVerboseEnv         = "SCW_VERBOSE_API"
+	cliDebugEnv           = "DEBUG"
+	cliNoCheckVersionEnv  = "SCW_NOCHECKVERSION"
+	cliTestWithRealAPIEnv = "TEST_WITH_REAL_API"
+	cliSecureExecEnv      = "SCW_SECURE_EXEC"
+	cliGatewayEnv         = "SCW_GATEWAY"
+	cliSensitiveEnv       = "SCW_SENSITIVE"
+	cliAccountApiEnv      = "SCW_ACCOUNT_API"
+	cliMetadataAPIEnv     = "SCW_METADATA_API"
+	cliMarketPlaceAPIEnv  = "SCW_MARKETPLACE_API"
+	cliComputePar1APIEnv  = "SCW_COMPUTE_PAR1_API"
+	cliComputeAms1APIEnv  = "SCW_COMPUTE_AMS1_API"
+	cliCommercialTypeEnv  = "SCW_COMMERCIAL_TYPE"
+	cliTargetArchEnv      = "SCW_TARGET_ARCH"
+)
+
+type Config interface {
+	GetAccessKey() string
+	GetSecretKey() string
+	GetApiUrl() string
+	GetInsecure() bool
+	GetDefaultOrganizationId() string
+	GetDefaultRegion() string
+	GetDefaultZone() string
+}
+
+type configV2 struct {
+	profile
+	ActiveProfile *string             `yaml:"active_profile"`
+	Profiles      map[string]*profile `yaml:"profiles"`
+
+	withProfile string
+}
+
+type profile struct {
+	AccessKey             *string `yaml:"access_key"`
+	SecretKey             *string `yaml:"secret_key"`
+	ApiUrl                *string `yaml:"api_url"`
+	Insecure              *bool   `yaml:"insecure"`
+	DefaultOrganizationId *string `yaml:"default_region"`
+	DefaultRegion         *string `yaml:"default_zone"`
+	DefaultZone           *string `yaml:"default_organization_id"`
+}
+
+func unmarshalConfV2(content []byte) (*configV2, error) {
+	var config configV2
+
+	err := yaml.Unmarshal(content, &config)
+	if err != nil {
+		return nil, err
+	}
+	return &config, nil
+
+}
+
+func (c *configV2) catchInvalidProfile() (*configV2, error) {
+	activeProfile, err := c.getActiveProfile()
+	if err != nil {
+		return nil, err
+	}
+
+	_, exist := c.Profiles[activeProfile]
+	if !exist {
+		return nil, fmt.Errorf("profile %s does not exist %s", activeProfile, inConfigFile())
+	}
+	return c, nil
+}
+
+func (c *configV2) getActiveProfile() (string, error) {
+	switch {
+	case c.withProfile != "":
+		return c.withProfile, nil
+	case os.Getenv(scwActiveProfileEnv) != "":
+		return os.Getenv(scwActiveProfileEnv), nil
+	case c.ActiveProfile != nil:
+		if *c.ActiveProfile == "" {
+			return "", fmt.Errorf("empty active profile %s", inConfigFile())
+		}
+		return *c.ActiveProfile, nil
+	default:
+		return "", nil
+	}
+}
+func (c *configV2) GetAccessKey() string {
+	// STEP 1: getenv
+	accessKey := getenv(scwAccessKeyEnv, terraformAccessKeyEnv)
+	if accessKey != "" {
+		return accessKey
+	}
+
+	// STEP 2: active profile
+	activeProfile, _ := c.getActiveProfile()
```

Comment:
Why is there no error check here?

Reply:
Those getters should not return error. Errors are caught on the `Load` with `catchInvalidProfile `

Reply:
Okay